src/confetti.js

```diff
@@ -64,8 +64,8 @@ export default class Confetti {
    *   The particle options.
    */
   start(opts = {}) {
-    if (!this.canvas || opts.canvasId !== this.canvasId) {
-      this.canvas = new Canvas(opts.canvasId);
+    if (!this.canvas || opts.canvasId !== this.canvasId || opts.canvasElement) {
+      this.canvas = new Canvas(opts.canvasId, opts.canvasElement);
```

Comment:
I can't quite remember what else `this.canvasId` is used for, but looks like it's also used to determine when to update. Do we need to account for that?

https://github.com/alexandermendes/vue-confetti/blob/854c4c1425883058456ab5a71f74440993c457fc/src/confetti.js#L100-L113

In fact, I guess instead of changing the canvas constructor we could do something like:

```js
/**
 * Get the canvas id from the given options.
 * @param {Object} opts
 *   The options.
 */
getCanvasId(opts) {
  const { canvasElement, canvasId } = opts;

  if (canvasElement && !(canvasElement instanceof HTMLCanvasElement)) {
    throw new Error('Element is not a valid HTMLCanvasElement');
  }

  return (canvasElement || {}).id || canvasId;
}

/**
 * Start dropping confetti.
 * @param {Object} opts
 *   The particle options.
 */
start(opts = {}) {
  const canvasId = this.getCanvasId(opts);

  if (!this.canvas || canvasId !== this.canvasId) {
    this.canvas = new Canvas(canvasId);
    this.canvasId = canvasId;
  }

  if (this.animationId) {
    cancelAnimationFrame(this.animationId); // Cancel any previous loop
  }

  this.createParticles(opts);
  this.canvas.updateDimensions();
  this.setParticlesPerFrame(opts);
  this.animationId = requestAnimationFrame(this.mainLoop.bind(this));
}
```

And then use `getCanvasId` for the update function too.

Reply:
Oops, the above slightly missed the point of passing the element directly. Still, some tidy up taking these ideas into account would be good.

Reply:
Also `start()` should account for `canvasElement` passed in via `opts`:
```js
  start(opts = {}) {
    const canvasId = this.getCanvasId(opts);

    if (!this.canvas || canvasId !== this.canvasId || opts.canvasElement) {
      this.canvas = new Canvas(canvasId, opts.canvasElement);
      this.canvasId = canvasId;
    }

    if (this.animationId) {
      cancelAnimationFrame(this.animationId); // Cancel any previous loop
    }

    this.createParticles(opts);
    this.canvas.updateDimensions();
    this.setParticlesPerFrame(opts);
    this.animationId = requestAnimationFrame(this.mainLoop.bind(this));
  }
```